e2e/testcases/override_role_refs_test.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/pkg/errors"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/reconcilermanager/controllers"
+	"kpt.dev/configsync/pkg/testing/fake"
+)
+
+func TestRootSyncRoleRefs(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.OverrideAPI, ntopts.Unstructured,
+		ntopts.RootRepo("sync-a"), // will declare namespace-a explicitly
+	)
+	rootSyncA := nomostest.RootSyncObjectV1Beta1FromRootRepo(nt, "sync-a")
+	syncAReconcilerName := core.RootReconcilerName(rootSyncA.Name)
+	syncANN := types.NamespacedName{
+		Name:      rootSyncA.Name,
+		Namespace: rootSyncA.Namespace,
+	}
+	if err := nt.Validate(controllers.RootSyncLegacyClusterRoleBindingName, "", &rbacv1.ClusterRoleBinding{},
+		testpredicates.ClusterRoleBindingHasSubjects(nomostest.DefaultRootReconcilerName, syncAReconcilerName)); err != nil {
+		nt.T.Fatal(err)
+	}
+	// Set some custom roleRef overrides
+	rootSyncA.Spec.SafeOverride().RoleRefs = []v1beta1.RootSyncRoleRef{
+		{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: "Role",
+				Name: "foo-role",
+			},
+			Namespace: "foo",
+		},
+		{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: "ClusterRole",
+				Name: "foo-role",
+			},
+		},
+		{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: "ClusterRole",
+				Name: "bar-role",
+			},
+		},
+		{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: "ClusterRole",
+				Name: "foo-role",
+			},
+			Namespace: "foo",
+		},
+	}
+	roleObject := fake.RoleObject(core.Name("foo-role"), core.Namespace("foo"))
+	clusterRoleObject := fake.ClusterRoleObject(core.Name("foo-role"))
+	clusterRoleObject.Rules = []rbacv1.PolicyRule{
```

Comment:
kinda weird having this in a specific test when these objects are defined in the test setup. Will we need this for other tests later? Or only for RootSync tests validating custom roleRefs?

Reply:
The reason this exists is because of the "safety" ClusterRole/Namespace that get added to all RootSync repositories by the test framework. Normally the RootSync gets cluster-admin by default and can manage ClusterRoles/Namespaces, but here the RootSync would error because it doesn't have permission to manage these resources.

For now this is the only test that adjusts the `roleRefs` field for `RootSyncs`, but yes if we added more tests that set `roleRefs` then we may need to reuse this `ClusterRole`.